src/Common/Auth/OneTimeAuth.php

```diff
@@ -0,0 +1,340 @@
+use OpenEMR\Common\Session\SessionUtil;
+use OpenEMR\Common\Utils\RandomGenUtils;
+
+class OneTimeAuth
+{
+    private $scope;
+    private $context;
+
+    public function __construct($context = 'portal', $scope = 'redirect')
+    {
+        // scope = portal/service tasks (reset, register). context = portal, patient etc.
+        $this->scope = $scope;
+        $this->context = $context;
+    }
+
+    /**
+     * @param array $p
+     * @return array|bool
+     * @throws \Exception
+     *
+     *   $p[
+     *    'pid' => '', // required for most onetime auth
+     *   'target_link' => '', // Onetime endpoint
+     *   'redirect_link' => '', // Where to redirect the user after auth
+     *   'enabled_datetime' => 'NOW', // Use a datetime if wish to enable for a future date.
+     *   'expiry_interval' => 'PT15M', // Always PTxx{Sec,Min,Day} PeriodTime
+     *   'email' => '']
+     */
+    public function createPortalOneTime(array $p = []): array|bool
+    {
+        $redirect_token = '';
+        $passed_in_pid = $p['pid'] ?? 0;
+        $valid = $this->isValidPortalPatient($passed_in_pid);
+        if (empty($valid['valid'] ?? null) || empty($passed_in_pid)) {
+            throw new \RuntimeException(xlt("Invalid Pid or patient not found!"));
+        }
+        $email = ($valid['email'] ?? '') ?: ($p['email'] ?? '');
+        $date_base = ($p['enabled_datetime'] ?? null) ?: 'NOW';
+        $expiry = new DateTime($date_base);
+        $expiry->add(new DateInterval($p['expiry_interval'] ?? 'PT15M'));
+        $token_raw = RandomGenUtils::createUniqueToken(32);
+        $pin = substr(str_shuffle(str_shuffle("0123456789")), 0, 6);
+        $token_encrypt = (new CryptoGen())->encryptStandard($token_raw);
+        if (empty($p['pid']) || empty($token_raw)) {
+            (new SystemLogger())->error("Onetime failed missing PID or token creation failed");
+            return false;
+        }
+
+        $redirect_raw = trim($p['redirect_link'] ?? '');
+        if (!empty($redirect_raw)) {
+            $redirect_plus = js_escape(['pid' => $passed_in_pid, 'to' => $redirect_raw]);
+            $redirect_token = (new CryptoGen())->encryptStandard($redirect_plus);
+            if (empty($redirect_token)) {
+                (new SystemLogger())->error("Onetime redirect failed encryption.");
+            }
+        }
+        if (!empty($p['target_link'] ?? null)) {
+            $site_addr = trim($p['target_link']);
+        } elseif ($this->context == 'portal') {
+            $site_addr = trim($GLOBALS['portal_onsite_two_address']);
+        }
+
+        $rtn['encoded_link'] = $this->encodeLink($site_addr, $token_encrypt, $passed_in_pid, $redirect_token);
+        $rtn['onetime_token'] = $token_encrypt;
+        $rtn['redirect_token'] = $redirect_token;
+        $rtn['pin'] = $pin;
+        $rtn['email'] = $email;
+
+        $save = $this->insertOnetime($passed_in_pid, $pin, $token_raw, $redirect_raw, $redirect_token, $expiry->format('U'));
```

Comment:
Does the redirect_token (encrypted) need to be stored? Would be best if this was not stored and the info was instead collected (and perhaps comfirmed with stored redirect_raw) when decrypting the token.

Reply:
agreed